### ISSUE TITLE
Migration Primer

### ISSUE THREAD

**Issue by blaggacao** [xoe-labs/dodoo-migrator#3]
@StefanRijnhart @pedrobaeza ... this is a question for migration cracks:

I've noticed that there are certain transitions that cannot be (atomically) done from within the module scope (like renaming or module deletion). As a response, I introduce a [Migration Primer](https://github.com/xoe-labs/click-odoo-migrator/blob/e5e7037f89845484a43e4702dc5f3a91b5eacf51/src/modules/parser.py#L18). Note that there is no need for a Migration Finalizer, as every "surviving" module can register it's own `end-*.py`., is there? (cc @guewen for marabunta finalizer - `post` - design)

Now the problem is that the statefulness of this primer spans several modules. Or expressed in other words: It depends on an implicit repository state. Problem is, this repository state does not have a naming, less semantic versioning, yet.

This lack of repository state is probably the reason why repository refactoring has been so hard (until now).

Question is: what can be done? Introduce a repository state snaphotting modules states? ~Any possibility to federate Migration Primer in a way so that every module passes it's own component dependent on the module state (module version)?~ (Will work for renaming, but not for deletes. Then follow-up question: are there really any true deletes in any continuous production db migration history??)

Just wanted to open up discussion to see if, together, we can come up with a better solution...

**Comment by blaggacao** [comment]
Further thinking: If we don't allow plain deletes as a valid transition, then probably any module could just cast a primer, i.e. `start-*.py` through a custom `MigrationManager` and a patch to odoo upstream.

And if we still need true deletes, we could lift them onto the controller plain like this `click-odoo-migrator` while providing external state ...

That might elegantly solve the repo state problem, a casted primer will scope it's own state just concerning the modules involved... Puh! I was already thinking this could get a *hard* reason to never decouple OCA repos...

**Comment by pedrobaeza** [comment]
I haven't understood most of your words, maybe the terminology or the long size, but I don't think there's any problem: we rename on base module. No more problems.

**Comment by blaggacao** [comment]
But @pedrobaeza, it's not necessary to explicitly say, that you don't have first class access to base in every use case.

Even in "your" use case: take any OCA repo.

About terminology: I'm trying to get a grasp on the problem in an abstract way. In order to find an abstract (and most suitable) solution.

**Comment by pedrobaeza** [comment]
Well, any other solution will require some kind of hook and hot-patching in current Odoo migration machinery. If you have an idea, make the PR over OpenUpgrade.

**Comment by blaggacao** [comment]
@pedrobaeza Indeed, it will. The relevant Odoo code is very ugly and old. I don't see hot patching as too much of a problem.
I've evaluated OpenUpgrade and unfortunately have come to the conclusion that it doesn't fit my requirement (I don't need a Odoo version packed, I just want a server API extension). That's why I do this click-odoo-migrator.

But back to the drawing board first, do you have any opinion on my questions?

**Comment by pedrobaeza** [comment]
What are they? Can you summarize them shortly?

**Comment by blaggacao** [comment]
Btw, there was another problem with base: I can't control/infere in it's version state (versioning is done by Editor). So even if I wanted, I could not hack something into base.

**Comment by blaggacao** [comment]
@pedrobaeza: most importantly: can true module deletes be declared "illegal" in "real" use cases. This would semantically solve most of the problem with a little patch to Odoo upstream.

I tend to conclude: they can.

**Comment by pedrobaeza** [comment]
You don't need to control base module version: you just need to check odoo version. If they put `1.1`, full version will be `11.0.1.1`, and you will know `11.0` part. That's the one you should control. About the illegal thing, I don't get your point.

**Comment by blaggacao** [comment]
@pedrobaeza let me give you a more realistic use case:
I have a thematic repo, and in the middle of v11 I realize that I need to completely refactor because Odoo comes up with some crappy thing (IOT) which invalidates most of my previous assumptions.
I want to migrate without being bound to the lifecycle of (implicit) base versioning.

With deletes being "illegal", I mean if it is a valid use cases that you would ever just 'drop' a module in a production database without any other module taking over part of the functionality.

Because if any module can cast a start script at the beginning of the migration, any module can organize for the refactoring of the modules involved based on their own federated versioning (lifecycle).

True deletes are the only operation that cannot be cleanly done atomically (with a single run and code state), because not having base, it would require a bare module of itself just for the purpose of de-registering before complete removal.

**Comment by pedrobaeza** [comment]
But you can already do that on your thematic repo. Nothing prevents you from deleting a module, merge it or so in the migration scripts of other modules. On the log, you will have an initial missing module error/warning, but at the end, everything will be correct. Look to this migration script that does exactly that: https://github.com/OCA/bank-payment/blob/9.0/account_payment_order/migrations/9.0.1.0.0/pre-migration.py

**Comment by blaggacao** [comment]
@pedrobaeza true, except for when the other module depends on the deleted module. Or am I missing something?

But thinking over this idea, then maybe there is only a very narrow use case for such primer / repo state migration:
- Either remove a module (without replacement)
- Or merge with a formerly dependent module (to adjust deps before migration)

Would you more or less agree?

**Comment by blaggacao** [comment]
Also I've seen in those famous Odoo migration scripts new module seeding and force installing as if there was a use case for it. I'm not sure if there is.

If I can conclude that I can avoid the primer, I'd be happy to leave that out because of the problems with their meta lifecycle.

**Comment by pedrobaeza** [comment]
OK, but this kind of things is something out of the scope of OpenUpgrade and OCA, and I only work inside that frame.

**Comment by blaggacao** [comment]
@pedrobaeza Ok, I appreciate the conversation. Thanks! When you're ready to innovate outside of that frame or broaden the use cases, I'm happy to discuss...

**Comment by pedrobaeza** [comment]
It's not a question of innovation, but principles, and I do everything open and inside OCA.

**Comment by blaggacao** [comment]
You see: I read "open" and "inside" as antagonists in your phrase. I'm not against the OCA, I'm not the hardest fan neither: 
- Still, I'm open. ;-)
- This is intended to be an open discussion (I still don't know the ultimate answer to my design dilemma, yet)

**Comment by pedrobaeza** [comment]
They are not antagonists, as everything in OCA is open, but what I don't want is to spread my efforts over other places that are not OCA or my company ones. If we collaborate, it's inside OCA framework. I can make exceptions on this if it's out of the OCA scope (for example, Python libraries), but it's not the case. 

And in this case, the migration path is clear for me: OpenUpgrade and everything around it. No other tools needed, and more if they are outside OCA.

**Comment by blaggacao** [comment]
@pedrobaeza Ok, I understand. But you cannot argue it's open (like in no limits). It's constraint to OCA. That's why to me those word were used are antagonists.

But ok, let's put it the other way round:
- I'd like to publish this in OCA (give me repo access, and I'll do it right now)
- Would that change your stance?

And itsn't it unlikely to be accepted because it goes against the consolidation of efforts around a topic (aka OpenUpgrade)?

If that last suspicion holds true, that's detrimental for innovation. So yes, it's about open and free innovation.

Don't you think I have a point here?

**Comment by pedrobaeza** [comment]
@blaggacao the problem is not the innovation, but the duplication of efforts and the waste of time. If you find a problem in current OpenUpgrade, openupgradelib or Odoo migration machinery, exposes or try to solve it by the known methods (PR or issue), but don't reinvent the wheel in that case.

And doing the PR or the issue should be something straight for the reviewers of the PR or commenters of the issue. Longer texts don't mean better explanations. Without the spirit of bothering you, but you ramble a lot, and this exhausts people to try to understand you. Look as some PRs you have made that are focused have been merged (the logo on OpenUpgrade, the taxes mapping on PoS, etc).

I remember the case also when you put an issue about merging modules, but your dissertation confused me a lot. Finally, we implemented something similar on openupgradelib.update_module_names by our own roots, and then I remembered that you propose that. If you would go straight to the point, we would find an arrangement sooner.

**Comment by blaggacao** [comment]
@pedrobaeza If I would like to refactor OCA upgrade machinery (towards a Server API extension), were I to succeed within reasonable time?

Innovation is about fast iterations. So yes, on the foundation of everything you say: It's about innovation, fast prototyping, sharing ideas, celebrating different perspectives, learn new things, and a welcoming community (you can't blame the pro-activity of good intentions for bad reactions).

And as it stands, for me David Arnold, OCA is not a place for new ideas to thrive. Not at all! I have my points of comparation, which drive my expectations. If I engage somewhere, it's not only because of the consolidated marketing power towards end users, but also because it's fun.

In conclusion, if there is no cross fertilization, which seems to be against your  stated principles, this `click-odoo-migrator` will simply strive to become the better alternative to OCA upgrade machinery and be the officially supported migration machinery for the Odoo-Operator. OCA people can choose to boycott, but this would be an even more extreme case of "closeness" vs "openness": not appealing to anybody.

As I say, at this moment, I don't have any problem to incorporate it into OCA, just provide my with a place for it to live and you will be able to unleash your collaboration.

**Comment by pedrobaeza** [comment]
I'm not going to boycott anything, just simply I won't spend efforts on that as I have all my needs solved with OCA upgrade machinery.

**Comment by blaggacao** [comment]
But do I understand correctly, that you also don't encourage to supply a place where this can live under OCA?

Btw, apart from politics: The initial problem is answered by those lines: https://github.com/odoo/odoo/blob/1dcddbd022d3de6f1383e014d3ea0308640b6922/odoo/modules/loading.py#L403-L426

Instead of casting something to the `start-` it just iterates over modules in case of "broken" dependencies caused by a removed module until the relevant pre-hook could kick in. This has the same effect as the theoretizised casted `start-` script.

Therefore we even don't need any primer, as I tend to think we can declare plain module removes illegal in a stable migration paths.

It can be that someone want's to remove a module, but it is extremely unlikely that this would be an operations of a _stable_ migration path.

For exceptional module deletes, we can add `delete` to the `click-odoo-migrator` yaml API.
